Add previous-scale annotation for idled resources

DirectXMan12 opened the pull request:
This records the previous scale of an idled resource on idling.
This is useful in situations where finding and loading the associated
endpoints object would non-ideal (e.g. the web console).

DirectXMan12 commented:
cc @smarterclayton @jwforres

knobunc commented:
LGTM

sjenning commented:
lgtm

ncdc commented:
@smarterclayton who is doing the api review? you / @deads2k / @liggitt / @jwforres ?

deads2k commented on pkg/unidling/api/types.go:
```diff
@@ -9,6 +9,10 @@ const (
 
+	// PreviousScaleAnnotation contains the previous scale of a scalable object
+	// (currently only applied by the idler)
+	PreviousScaleAnnotation = "scale.openshift.io/previous-scale"
```
This name doesn't match what this is doing.  When I scale an object, I don't get this value filled in.  This looks tightly coupled to idling.

ncdc replied:
Yeah, I would probably call this idling.alpha.openshift.io/previous-scale or something like that

DirectXMan12 replied:
I can, but when I was talking to @smarterclayton, we were discussing the idea of making something like this automatically applied on scales in the future (and then have scale.openshift.io/xyz passed through the scale subresource so that it can act like a prototype API field, just like other objects), so using something like this would future-proof that somewhat.

DirectXMan12 replied:
(although using `scale.openshift.io` instead of `scale.kubernetes.io` might complicate that somewhat)

deads2k replied:
> I can, but when I was talking to @smarterclayton, we were discussing the idea of making something like this automatically applied on scales in the future (and then have scale.openshift.io/xyz passed through the scale subresource so that it can act like a prototype API field, just like other objects), so using something like this would future-proof that somewhat.

I'm not really convinced that I'd like to use an annotation to remember previous state (see problems with the other kube attempts to use annotations to store history).  I'm also not seeing a lot of value for "undo" in the general case.  It seems like it would open us up to a new class of privilege escalation attacks if we had a controller that attempted to reconcile this annotation automatically.

I'd rather target the information that we want to preserve for its current usage.

@smarterclayton is there some more complete design that covers how we'd resolve the problems of history tracking annotations and privilege escalation attacks based on controllers watching annotations (think DC scaling existing RC, not scaling the current object)?  Are you ok with scoping this to address our current, narrow need first?

deads2k commented:
Name needs updating and it would be nice to see a CLI usage demonstrating your intent, which I think is something like:
 1. create thing
 2. scale thing up
 3. idle thing
 4. knowing only the thing.Name, scale it back to the value from step 2.

Other than the name, I'm fine with this.

DirectXMan12 commented:
> CLI usage

ack, will add tomorrow

deads2k commented:
> ack, will add tomorrow

Thanks.  There's probably a `test/cmd/*.sh` script that would fit.

smarterclayton commented:
Edit on a resource is edit on scale, so what's the privilege escalation?

Preserving the user's previous scale value is important - if the user has
an idled pet set, restoring scale to 1 on unidle is completely wrong,
likewise an app that is supposed to be HA (2 pods) that gets idled back to
1, but no human notices.

On Tue, Aug 16, 2016 at 5:18 PM, David Eads <notifications@github.com>
wrote:

> ack, will add tomorrow
>
> Thanks. There's probably a test/cmd/*.sh script that would fit.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/10421#issuecomment-240242255>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p89IQC-PXl_Hw_ft_SOtFtdzwdrNks5qgikfgaJpZM4JknXz>
> .
>

deads2k commented:
> Preserving the user's previous scale value is important - if the user has
an idled pet set, restoring scale to 1 on unidle is completely wrong,

For idling, sure.  No issue with this for idling.  In the general case though, I don't see why I would try to build a scaling history

smarterclayton commented:
Maybe I missed the context, but some status displaying recent scaling might
actually be useful for automated controllers.  Hard to say, knowing "why"
someone last scaled (human, or automated) might be relevant.  And is
relevant for idling in some contexts.

On Tue, Aug 16, 2016 at 6:02 PM, David Eads <notifications@github.com>
wrote:

> Preserving the user's previous scale value is important - if the user has
> an idled pet set, restoring scale to 1 on unidle is completely wrong,
>
> For idling, sure. No issue with this for idling. In the general case
> though, I don't see why I would try to build a scaling history
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/10421#issuecomment-240253707>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p96lK1xzvY3f9NPO4gF78suoRU7Zks5qgjNxgaJpZM4JknXz>
> .
>

deads2k commented:
> Maybe I missed the context, but some status displaying recent scaling might
actually be useful for automated controllers.

Maybe, but I'd rather add it for a concrete reason, set in a way that makes sense, and get in a place with a name that makes sense.  This pull does none of those things. 

To be clear, we're arguing over a name here, not the use in this narrow case. The annotation is set only by idling for a narrow reason, I think it deserves a narrowly scoped name.

deads2k commented:
Also, scale on a DC does not edit the DC and without access to the rcs you can never reproduce the effect since you don't know the shape of the RC you're scaling by touching the DC.  It's like a dc status.  It's actually mucking directly with rcs.

DirectXMan12 commented:
> Also, scale on a DC does not edit the DC 

I believe it does, unless someone changed it since I added support for the `/scale` subresource on DCs.  `oc scale` didn't used to, but when I added the scale endpoint, we changed it to edit the DC and reconcile the RCs with the DCs, so unless someone changed it after that, the `/scale` subresource modifies DCs.

smarterclayton commented:
Scale on a DC does edit the DC

DirectXMan12 commented:
> To be generally useful, you need to have the contract on the polymorphic scale endpoint change the annotation. This pull does none of those things. 

Sure, that's a thing that requires an update in the future (and it's something I intend to work on, as it would be generally useful to be able to prototype scale API fields the same way we prototype the addition of fields to other API objects), but I mainly wanted to have something that was forwards compatible, so to speak.

deads2k commented:
> but I mainly wanted to have something that was forwards compatible, so to speak.

Narrow name for a narrow case.  The general solution to a theoretical general problem that requires upstream API changes we won't pick for a general need we don't have is probably more than is worth picking up here.

deads2k commented:
> Narrow name for a narrow case. The general solution to a theoretical general problem that requires upstream API changes we won't pick for a general need we don't have is probably more than is worth picking up here.

"idling.alpha.openshift.io/previous-scale" is perfect.  It's scoped to its use, the name identifies who is setting it and it matches the behavior you'd expect when idling and unidling.

"scale.openshift.io/previous-scale" is overly broad.  It suggests that the scale operation is controlling this (it isn't).  To make it behave as expected, you'd have to change the contract on the polymorphic `/scale` endpoint.  Even if you did that, such an annotation wouldn't be scoped to openshift.  And, if you want this power out of scale, then retrieving via scale makes sense and this doesn't do that either.

ncdc commented:
We can always deprecate `idling.alpha.openshift.io/previous-scale` and replace it with something else that's more appropriate for a broader scope when the time comes. I suggest we just use `idling.alpha.openshift.io/previous-scale` for now.

DirectXMan12 commented:
yeah, for the sake of moving this forward I'll just change it to the annotation suggested by @deads2k

deads2k commented:
lgtm [merge]

DirectXMan12 commented:
Missing quote in the cmd test.  Fixed.

smarterclayton commented:
[merge]

deads2k commented:
@DirectXMan12 any real changes?

DirectXMan12 commented:
@deads2k no, just rebased b/c there was a weird error in the console log of the tests about one of the files needing a merge.

deads2k commented:
[merge]

ncdc commented:
[test] for fun
